Contractor management api

ckc-efehan opened the pull request:
This pull request implements a comprehensive API for contractor management within the Remsfal project. The API allows users to manage contractors within projects, including:

Listing all contractors for a project with pagination
Creating new contractors with company information
Retrieving detailed information about individual contractors
Updating contractor data
Deleting contractors

The implementation includes REST endpoints, controllers for business logic, and JSON data models for representing contractors and their employees. The API is fully integrated into the existing project structure and includes permission checks for all operations.

astanik commented:
@Mehdi030 You've only left one comment so far. The review is still pending.

Mehdi030 reviewed (approved):
Hello, this is what i think:

What I liked:

clean structure between controller, service and repository.
REST endpoints cover all basic contractor operations.
Permission checks already included – good to see.
Suggestions for improvement:

Add validation for input fields to avoid saving incomplete contractor data.
Make sure error responses use consistent and appropriate HTTP status codes.
Pagination is mentioned – if not fully implemented yet, consider adding parameters like page and limit.
Looks like a solid base overall – nice work!

astanik commented on remsfal-services/remsfal-platform/src/main/java/de/remsfal/service/control/ContractorController.java:
```diff
@@ -0,0 +1,198 @@
+                                               final Integer offset, final Integer limit) {
+        logger.infov("Retrieving contractors for project (id = {0})", projectId);
+        // Verify user has access to the project
+        verifyProjectAccess(user, projectId);
```
This authorization check should not take place in the business logic, but rather in the boundary. And this already happens in the ContractorController with checkReadPermissions(projectId). So you could simply remove this check.

astanik commented on remsfal-services/remsfal-platform/src/main/java/de/remsfal/service/control/ContractorController.java:
```diff
@@ -0,0 +1,198 @@
+     */
+    public long countContractors(final UserModel user, final String projectId) {
+        // Verify user has access to the project
+        verifyProjectAccess(user, projectId);
```
same here

astanik commented on remsfal-services/remsfal-platform/src/main/java/de/remsfal/service/control/ContractorController.java:
```diff
@@ -0,0 +1,198 @@
+    public ContractorModel getContractor(final UserModel user, final String projectId, final String contractorId) {
+        logger.infov("Retrieving contractor (id = {0}) for project (id = {1})", contractorId, projectId);
+        // Verify user has access to the project
+        verifyProjectAccess(user, projectId);
```
and here

astanik reviewed (changes requested):
You could also use immutables for the JSON classes to save code, but that's not necessary. I like the rest of it very much.

ckc-efehan commented:
You are absolutely right and I have implemented the change as suggested. I have removed the redundant authorization checks from the ContractorController and adjusted all tests to ensure they continue to pass successfully. The separation of concerns is now clearer, with the boundary layer (ContractorResource) being responsible for authorization and the controller layer focusing on business logic.

astanik commented on remsfal-services/remsfal-platform/src/main/java/de/remsfal/service/control/ProjectController.java:
```diff
@@ -68,6 +68,13 @@ public ProjectModel createProject(final UserModel user, final ProjectModel proje
 
     public ProjectModel getProject(final UserModel user, final String projectId) {
         logger.infov("Retrieving a project (id = {0})", projectId);
+        // First check if the project exists
+        ProjectEntity project = projectRepository.findById(projectId);
+        if (project == null) {
+            throw new NotFoundException("Project not found");
+        }
+
+        // Then check if the user is a member of the project
```
This is not clear to me. Why is this additional check required?
```suggestion
```

ckc-efehan replied:
The redundant authorization checks in the codebase were indeed unnecessary. It was simply my mistake to implement these duplicate checks.
